staging/src/k8s.io/api/core/v1/pod/util.go

```diff
@@ -0,0 +1,391 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```

Comment:
@pohly what do you think ?

Reply:
> I think is better to keep them as an internal detail.

The functions might be implementation specific and thus not worthy for staging, but what they test is part of the stable API - I guess one can argue both ways.

> a quick skim through it seems only the IsReady function is used

That's indeed not as much as I was [told earlier](https://github.com/kubernetes/kubernetes/pull/117379#issuecomment-1511818263).

@carlory: it probably makes more sense to duplicate `IsReady` in test/e2e/framework/pod.

Reply:
That list includes also things from `k8s.io/kubernetes/pkg/api/v1/pod`. We also would need to duplicate that inside test/e2e/framework/pod, like `VisitContainers`.

Reply:
https://grep.app/search?q=VisitContainers shows that there are plenty of copies of this function.

[staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go) uses it and had to make a copy in "k8s.io/kubectl/pkg/util/podutils".

The E2E framework can use that copy, but that doesn't feel right. "k8s.io/kubernetes/pkg/api/v1/pod" would be a better place.

@aojea: what do you prefer?

Reply:
@pohly 

kubectl has already copied those funcs, see

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/util/podutils/podutils.go#L192


Can  e2e framewwork import it instead.

Reply:
everthing in the podutils.go are copied from "k8s.io/kubernetes/pkg/api/v1/pod".

Reply:
why don't we copy too , we are goi

> The E2E framework can use that copy, but that doesn't feel right. "k8s.io/kubernetes/pkg/api/v1/pod" would be a better place.

I don't understand this comment, are we not already depending on `"k8s.io/kubernetes/pkg/api/v1/pod"`?

what are the options,
1. duplicate in test/e2e/framework/pod
2. import k8s.io/kubectl/pkg/util/podutils
3. import k8s.io/kubernetes/pkg/api/v1/pod

since that function is not likely to change it kinds of feel that 1. is the best option, is not that the compiled binary size is going to grow, no?

Reply:
> are we not already depending on "k8s.io/kubernetes/pkg/api/v1/pod"?

Yes, but that's not a suitable dependency for the E2E framework. It works, but it is conceptually unclean for a Kubernetes client (like the E2E test suite) to depend on internal code.

We can import "k8s.io/kubectl/pkg/util/podutils" (works out of the box, but feels weird) or we can import "k8s.io/kubernetes/pkg/api/v1/pod" (must add VisitContainers first).

I feel we are putting too much effort into this. Let's take the low road and:
- use `k8s.io/kubectl/pkg/util/podutils.VisitContainers` in the E2E framework
- duplicate all other code in `test/e2e/framework/pod`

Reply:
@pohly 

Why don't we consider using other functions in `k8s.io/kubectl/pkg/util/podutils`, such as `IsPodAvailable `、`IsPodReady ` ?

Reply:
If they are there, then let's use them. I hadn't checked that.